machine/v1beta1/types_vsphereprovider.go

```diff
@@ -87,12 +87,61 @@ type NetworkSpec struct {
 	Devices []NetworkDeviceSpec `json:"devices"`
 }
 
+// AddressesFromPool is an IPAddressPool that will be used to create
+// IPAddressClaims for fulfillment by an external controller.
+type AddressesFromPool struct {
+	// group of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Group string `json:"group"`
+	// resource of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// name of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+
 // NetworkDeviceSpec defines the network configuration for a virtual machine's
 // network device.
 type NetworkDeviceSpec struct {
-	// NetworkName is the name of the vSphere network to which the device
-	// will be connected.
-	NetworkName string `json:"networkName"`
+	// networkName is the name of the vSphere network or port group to which the network
+	// device will be connected.
+	// +optional
+	NetworkName string `json:"networkName,omitempty"`
+
+	// gateway4 is the IPv4 gateway used by this device. Required if an IPv4 IP address
+	// is provided in IPAddrs.
+	// +kubebuilder:validation:Format=ipv4
+	// +optional
+	Gateway4 string `json:"gateway4,omitempty"`
+
+	// gateway6 is the IPv6 gateway used by this device.
+	// +kubebuilder:validation:Format=ipv6
+	// +optional
+	Gateway6 string `json:"gateway6,omitempty"`
+
+	// ipAddrs is a list of one or more IPv4 and/or IPv6 addresses and CIDR to assign
+	// to this device. If ipAddrs is not defined, or does not contain an IPv4 and/or,
+	// IPv6 address DHCP will assign addresses for IPv4 and/or IPv6.
+	// +kubebuilder:example=192.168.1.100/24
+	// +kubebuilder:example=2001:DB8:0000:0000:244:17FF:FEB6:D37D/64
+	// +optional
+	IPAddrs []string `json:"ipAddrs,omitempty"`
+
+	// nameservers is a list of IPv4 and/or IPv6 addresses used as DNS
+	// nameservers. a nameserver is not provided by a fulfilled IPAddressClaim.
+	// if DHCP is not the source of IP addresses for this network device, nameservers
+	// should include a defined nameserver.
+	// +optional
+	Nameservers []string `json:"nameservers,omitempty"`
+
+	// addressesFromPools is a list of references to IP pool types and instances
+	// which are handled by an external controller.  The external controller
+	// will assign IP addresses in accordance with the IP pool instances.
+	AddressesFromPools []AddressesFromPool `json:"addressesFromPool,omitempty"`
```

Comment:
Is this required?

Reply:
Yes, this is used by the reconciler to construct an IPAddressClaim.  An external controller recognizes its known IP pool type instances included in the IPAddressClaim to fulfill the claim.

Reply:
Ok, so if this is required can you please add that marker, and explain how this interacts with other fields, eg, the IPAddrs field?

Not sure what the interaction between the two is but they look to be providing pretty similar information